File: Problems 1-9/p9.jl
"""
There exists exactly one Pythagorean triplet for which a + b + c = 1000.
    Find the product abc.
"""

unique([prod([x, y, z]) for (x, y) in Base.product(1:500, 1:500) for z in 1:500 if ((x^2 + y^2 == z^2) & (x + y + z == 1000))])

"""
julia> @btime unique([prod([x, y, z]) for (x, y) in Base.product(1:500, 1:500) for z in 1:500 if ((x^2 + y^2 == z^2) & (x + y + z == 1000))])
  41.436 ms (11 allocations: 912 bytes)
1-element Vector{Int64}:
 31875000

julia> @time unique([prod([x, y, z]) for (x, y) in Base.product(1:500, 1:500) for z in 1:500 if ((x^2 + y^2 == z^2) & (x + y + z == 1000))])
 0.120871 seconds (296.80 k allocations: 15.262 MiB, 65.11% compilation time)
1-element Vector{Int64}:
31875000
"""

File: Problems 1-9/p9.py
"""
There exists exactly one Pythagorean triplet for which a + b + c = 1000.
    Find the product abc.
"""
from math import prod
import itertools
from numpy import unique
from time import perf_counter

unique(
    [
        prod([x, y, z])
        for (x, y) in itertools.product(range(1, 501), range(1, 501))
        for z in range(1, 501)
        if ((x**2 + y**2 == z**2) and (x + y + z == 1000))
    ]
)[0]


def main():
    t1 = perf_counter()
    print(
        unique(
            [
                prod([x, y, z])
                for (x, y) in itertools.product(range(1, 501), range(1, 501))
                for z in range(1, 501)
                if ((x**2 + y**2 == z**2) and (x + y + z == 1000))
            ]
        )[0]
    )
    t2 = perf_counter()
    return t2 - t1
